File: js/AccesoController.js
// Generar un salt aleatorio utilizando la clase Crypto.getRandomValues()
function generarSalt() {
    var randomBytes = new Uint8Array(32);
    crypto.getRandomValues(randomBytes);
    return btoa(String.fromCharCode.apply(null, randomBytes));
  }
  
  // Convertir una cadena a hash SHA256
  function convertirSha256(cadena, salt) {
    var encoder = new TextEncoder();
    var data = encoder.encode(cadena + salt);
    return crypto.subtle.digest("SHA-256", data)
      .then(function (hashBytes) {
        var hashArray = Array.from(new Uint8Array(hashBytes));
        return hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
      });
  }
  
  // GET: Acceso
  function login() {
    // Renderizar la vista correspondiente
  }
  
  function registrar() {
    // Renderizar la vista correspondiente
  }
  
  // Registrar un usuario en la base de datos
  function registrarUsuario(oUsuario) {
    var registrado;
    var mensaje;
  
    if (oUsuario.Clave == oUsuario.ConfirmarClave) {
      var salt = generarSalt(); // Generar un salt aleatorio
      convertirSha256(oUsuario.Clave, salt).then(function (hash) {
        oUsuario.Clave = hash;
  
        // Guardar el usuario en la base de datos
        var params = new URLSearchParams();
        params.append('Correo', oUsuario.Correo);
        params.append('Clave', oUsuario.Clave);
        params.append('Salt', salt);
        fetch('/api/registrarUsuario', {
          method: 'POST',
          body: params
        }).then(function (response) {
          return response.json();
        }).then(function (data) {
          registrado = data.registrado;
          mensaje = data.mensaje;
          document.getElementById("mensaje").innerText = mensaje;
          if (registrado) {
            window.location.href = "/Acceso/Login";
          }
        });
      });
    } else {
      mensaje = "Las contraseñas no coinciden";
      document.getElementById("mensaje").innerText = mensaje;
    }
  }
  
  // Validar las credenciales de un usuario en la base de datos
  function validarUsuario(oUsuario) {
    convertirSha256(oUsuario.Clave, oUsuario.Salt).then(function (hash) {
      oUsuario.Clave = hash;
  
      // Validar las credenciales del usuario
      var params = new URLSearchParams();
      params.append('Correo', oUsuario.Correo);
      params.append('Clave', oUsuario.Clave);
      fetch('/api/validarUsuario', {
        method: 'POST',
        body: params
      }).then(function (response) {
        return response.json();
      }).then(function (data) {
        var idUsuario = data.idUsuario;
        if (idUsuario != 0) {
          oUsuario.IdUsuario = idUsuario;
          sessionStorage.setItem("usuario", JSON.stringify(oUsuario));
          window.location.href = "/Home/Index";
        } else {
          document.getElementById("mensaje").innerText = "usuario no encontrado";
        }
      });
    });
  }

File: Controllers/AccesoController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Web;
using System.Web.Mvc;

using PRUEBAS_LOGIN.Models;

using System.Data.SqlClient;
using System.Data;
using PRUEBAS_LOGIN.Utils;

namespace PRUEBAS_LOGIN.Controllers
{
    public class AccesoController : Controller
    {

        static string cadena = "Data Source=DESKTOP-9B4FHOH\\SQLEXPRESS;Initial Catalog=DB_ACCESO;Trusted_Connection=False; User ID=sa;Password=123456";



        // GET: Acceso
        public ActionResult Login()
        {
            return View();
        }


        public ActionResult Registrar()
        {
            return View();
        }


        [HttpPost]
        public ActionResult Registrar(Usuario oUsuario)
        {
            bool registrado;
            string mensaje;
            Security _security = new Security();

            if (oUsuario.Clave == oUsuario.ConfirmarClave)
            {

                string salt = _security.GenerateSalt(); // Generar un salt aleatorio
                oUsuario.Clave = _security.ConverToSha256(oUsuario.Clave, salt); // Agregar el salt a la contraseña y generar el hash SHA256

                using (SqlConnection cn = new SqlConnection(cadena))
                {

                    SqlCommand cmd = new SqlCommand("sp_RegistrarUsuario", cn);
                    cmd.Parameters.AddWithValue("Correo", oUsuario.Correo);
                    cmd.Parameters.AddWithValue("Clave", oUsuario.Clave);
                    cmd.Parameters.AddWithValue("Salt", salt); // Guardar el salt en la base de datos
                    cmd.Parameters.Add("Registrado", SqlDbType.Bit).Direction = ParameterDirection.Output;
                    cmd.Parameters.Add("Mensaje", SqlDbType.VarChar, 100).Direction = ParameterDirection.Output;
                    cmd.CommandType = CommandType.StoredProcedure;

                    cn.Open();

                    cmd.ExecuteNonQuery();

                    registrado = Convert.ToBoolean(cmd.Parameters["Registrado"].Value);
                    mensaje = cmd.Parameters["Mensaje"].Value.ToString();


                }

                ViewData["Mensaje"] = mensaje;

                if (registrado)
                {
                    return RedirectToAction("Login", "Acceso");
                }
                else
                {
                    return View();
                }

            }
            else
            {
                ViewData["Mensaje"] = "Las contraseñas no coinciden";
                return View();
            }

        }

        [HttpPost]
        public ActionResult Login(Usuario oUsuario)
        {
            Security _security = new Security();
            using (SqlConnection cn = new SqlConnection(cadena))
            {
                cn.Open();

                // Obtener el salt de la base de datos correspondiente al correo electrónico del usuario
                SqlCommand saltCmd = new SqlCommand("SELECT Salt FROM Usuario WHERE Correo = @Correo", cn);
                saltCmd.Parameters.AddWithValue("Correo", oUsuario.Correo);
                string salt = saltCmd.ExecuteScalar().ToString();

                // Generar el hash SHA256 de la contraseña del usuario junto con el salt recuperado
                oUsuario.Clave = _security.ConverToSha256(oUsuario.Clave, salt);

                // Validar las credenciales del usuario
                SqlCommand cmd = new SqlCommand("sp_ValidarUsuario", cn);
                cmd.Parameters.AddWithValue("Correo", oUsuario.Correo);
                cmd.Parameters.AddWithValue("Clave", oUsuario.Clave);
                cmd.CommandType = CommandType.StoredProcedure;

                oUsuario.IdUsuario = Convert.ToInt32(cmd.ExecuteScalar().ToString());

            }

            if (oUsuario.IdUsuario != 0)
            {

                Session["usuario"] = oUsuario;
                return RedirectToAction("Index", "Home");
            }
            else
            {
                ViewData["Mensaje"] = "usuario no encontrado";
                return View();
            }



        }

        // Generar un salt aleatorio utilizando la clase RNGCryptoServiceProvider


    }
}